### PR TITLE
LB policy API improvements

### PR DESCRIPTION
**Pull request by markdroth**
A few LB policy API improvements that I've been meaning to get to for a while now:
- Switch to using `absl::variant<>` instead of an enum to differentiate between the types of `PickResult`.
- Add an explicit separate result type for drops.  (Previously they were represented by "Complete" with the subchannel pointer being null.)
- The picker now sets the status to be used for drops, rather than hard-coding it in the channel.
- Replace `grpc_error_handle` with `absl::Status` in `PickResult` and in `recv_trailing_metadata_ready` callback.

### CODE REVIEW

**Comment by ctiller** [comment]
Would this be better as an `absl::variant<PickComplete, PickFailed, PickQueue, PickDrop>`?

It would allow us to be explicit (and compiler verified) about the fields that are active, and likely makes for a cleaner conversion path to the promise related stuff we've been tinkering with.

**Comment by markdroth** [comment]
Hmm.  I tried converting it to use `absl::variant<>`, which you can see here:

https://github.com/markdroth/grpc/commit/4ff04407afe5d6758e503857dc0191da4881a22d

However, I don't think I like it.  It seems to make the code more awkward.  Can you think of a way to make it cleaner?

**Comment by ctiller** [comment]
I left a comment on the linked change... not sure whether it will show up for you.

**Comment by markdroth** [comment]
Yeah, I saw the comment, thanks.  I'm gonna try the approach you suggested and see how it looks.

**Comment by markdroth** [comment]
I pushed another change to the other branch to try to clean up the use of `absl::variant<>`:

https://github.com/markdroth/grpc/commit/292c7f3525eb399a8bb8f38188dcf8f984979a7a

You can see the full diff between this branch and that one here:

https://github.com/markdroth/grpc/compare/lb_policy_api_improvements...markdroth:lb_policy_api_improvements.alt

It still seems like the code is overall a bit more cumbersome.  I do like the idea of having the compiler enforce that the right fields are set for each type of response, but I'm not sure if it's worth the code complexity.  After all, it doesn't seem that likely that LB policies will return the wrong set of fields.  And if they are going to have a problem with that, even using `absl::variant<>` won't really prevent all such problems (e.g., the LB policy can still return `CompletePick` without actually setting the subchannel).

WDYT?

**Comment by ctiller** [comment]
Honestly I think the variant version reads a whole lot better - I can far
more easily scan the code and find the piece that handles each case, and I
can (as a non-expert in load balancing) easily know which fields are needed
or not.

I don't think the bar for improvement should be 'addresses all known
problems', but if we wanted to do so, we could give CompletePick a
constructor that takes required fields, and add accessors, and make the
fields private.

**Comment by markdroth** [comment]
Okay.  I've brought the `absl::variant<>` changes into this branch.

File: src/core/ext/filters/client_channel/lb_policy.h [hunk 1]
```diff
@@ -419,6 +434,33 @@ class LoadBalancingPolicy : public InternallyRefCounted<LoadBalancingPolicy> {
   std::unique_ptr<ChannelControlHelper> channel_control_helper_;
 };
 
+//
+// Implementation
+//
+
+template <typename T>
```
Comment (ctiller) [inline src/core/ext/filters/client_channel/lb_policy.h:441]: I copy/pasted some fragments from the promises CL to make this:
https://godbolt.org/z/sWK3n7zcn

Suggest we move in the Overload type, and probably the associated Match function as a way of generalizing this part of the CL.

I can do so tomorrow if you'd like.

File: src/core/ext/filters/client_channel/lb_policy.cc [hunk 2]
```diff
@@ -91,6 +91,29 @@ LoadBalancingPolicy::UpdateArgs& LoadBalancingPolicy::UpdateArgs::operator=(
   return *this;
 }
 
+//
+// LoadBalancingPolicy::PickResult
+//
+
+LoadBalancingPolicy::PickResult LoadBalancingPolicy::PickResult::Complete(
```
Comment (ctiller) [inline src/core/ext/filters/client_channel/lb_policy.cc:98]: Question: do we still want these API's, given that we can now unambiguously construct a type with the right fields?

Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy.cc]: Good idea, done.

File: src/core/ext/filters/client_channel/lb_policy.h (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy.h:446]: Sure, if you want to throw together a PR to add this `Match()` function somewhere in src/core/lib/gprpp, we can use that here.

Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy.h:458]: Ping?  I'm happy to change this PR to use that sort of `Match()` function if you want to send me a PR to create it.

Comment (ctiller) [reply src/core/ext/filters/client_channel/lb_policy.h:458]: #26640

Comment (ctiller) [reply src/core/ext/filters/client_channel/lb_policy.h:458]: (sorry about the delay, detangling stuff from the promises PR got tricky)

Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy.h]: I've tried changing this to use the new `Match()` API, but the compiler is generating lock annotation errors.  It looks like something in the templating is breaking the lock annotation flow here.  Any idea how to fix this?

Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy.h]: As we discussed, I changed this back to using a custom helper function that doesn't break the lock annotation, but I've kept it out of the LB policy API, and I've added a TODO about changing it to use `Match()` once we find a way to not break the lock annotations.

**Comment by markdroth** [comment]
The "Bazel RBE Windows Debug C/C++" and "Basic Tests C# Windows" tests are infrastructure failures.

Known issues: #26595

**Review by ctiller — APPROVED** [review]

**Pull request merged**